pandas/core/indexes/base.py

```diff
@@ -2749,11 +2745,10 @@ def union(self, other):
                                   stacklevel=3)
 
         # for subclasses
-        return self._wrap_union_result(other, result)
+        return self._wrap_setop_result(other, result)
 
-    def _wrap_union_result(self, other, result):
-        name = self.name if self.name == other.name else None
-        return self.__class__(result, name=name)
+    def _wrap_setop_result(self, other, result):
+        return self._constructor(result, name=get_op_result_name(self, other))
```

Comment:
Would shallowcopy work here? Constructor isn’t used very often in the indexes.

Reply:
@jbrockmendel  repeating (part of) reply from a few months ago:
> the issue comes down to whether we use `self._shallow_copy()` to return the result of set operations. When I tried using `self._shallow_copy()` in `Index._wrap_setop_result()`, and looked at what was happening, then `CategoricalIndex` was the problem. So it was easier to leave the existing implementation of `Index._wrap_setop_result`. If we were to handle #10186, then I could experiment more with changing the implementation of `Index._wrap_setop_result`.